README.md
# Eureka Client for Rust

This project is currently in an **alpha state** as indicated by the version number.

## TODO

- [ ] async support
- [ ] auth support
- [ ] examples

### What Works

- Registering a service with Eureka
- Sending keep-alive heartbeats to Eureka
- Re-registering after Eureka restarting

### What is implemented but untested

- Making requests to services connected via Eureka

### What is not implemented

- DNS and AWS resolvers

## Installation

Add `eureka-client` to your `Cargo.toml` and add `extern crate eureka_client` to your project's root.

## Usage

To initialize a eureka client and register with eureka, you'll do something similar to this:

```rust
use eureka::{BaseConfig, EurekaClient, PortData};

pub fn init_eureka(
    server_host: String,
    server_port: u16,
    instance_ip_addr: String,
    instance_port: u16,
) -> EurekaClient {
    let mut config = BaseConfig::default();
    config.eureka.host = server_host;
    config.eureka.port = server_port;
    config.instance.ip_addr = instance_ip_addr;
    config.instance.port = Some(PortData::new(instance_port, true));
    let eureka = EurekaClient::new(config);
    eureka.start();
    eureka
}
```

You'll need to keep this client alive for as long as you intend to be connected to Eureka.
For example, in Rocket, you can manage it as state and access it via your routes as you would with other state, e.g. calling our above function:

```rust
rocket::ignite()
    .mount("/api/", routes![])
    .manage(init_eureka(
        dotenv!("EUREKA_HOST").into(),
        dotenv!("EUREKA_PORT")
            .parse()
            .expect("Eureka port not valid"),
        env::var("EUREKA_INSTANCE_IP").unwrap_or_else(|_| "127.0.0.1".to_string()),
        env::var("ROCKET_PORT")
            .map_err(|_| ())
            .and_then(|port| port.parse().map_err(|_| ()))
            .unwrap_or(8080),
    ))
```

This client registers with eureka by default. You can disable registration by setting `config.eureka.register_with_eureka = false`
if you just want to use this client to make requests.

## License

MIT

CHANGELOG.md
## 0.2.0

- Replace json with strong-xml, fix registration errors

## 0.0.4

- Bump dependencies

src/aws.rs
use std::collections::HashMap;

use reqwest::{Client, Response};
use serde_json::Value;

#[derive(Debug, Clone)]
pub struct AwsMetadata {
    client: Client,
    host: String,
}

impl AwsMetadata {
    pub fn new(config: &HashMap<String, Value>) -> Self {
        AwsMetadata {
            client: Client::new(),
            host: config
                .get("host")
                .map(|host| host.as_str().unwrap().to_string())
                .unwrap_or_else(|| String::from("169.254.169.254")),
        }
    }

    fn fetch_metadata(&self) -> HashMap<&'static str, String> {
        let mut results = HashMap::with_capacity(11);
        results.insert("ami-id", self.lookup_metadata_key("ami-id"));
        results.insert("instance-id", self.lookup_metadata_key("instance-id"));
        results.insert("instance-type", self.lookup_metadata_key("instance-type"));
        results.insert("local-ipv4", self.lookup_metadata_key("local-ipv4"));
        results.insert("local-hostname", self.lookup_metadata_key("local-hostname"));
        results.insert(
            "availability-zone",
            self.lookup_metadata_key("placement/availability-zone"),
        );
        results.insert(
            "public-hostname",
            self.lookup_metadata_key("public-hostname"),
        );
        results.insert("public-ipv4", self.lookup_metadata_key("public-ipv4"));
        results.insert("mac", self.lookup_metadata_key("mac"));
        results.insert(
            "accountId",
            self.lookup_instance_identity()
                .and_then(|i| i["accountId"].as_str().map(|id| id.to_owned())),
        );
        let mac = results["mac"].clone().unwrap();
        results.insert(
            "vpc-id",
            self.lookup_metadata_key(&format!("network/interfaces/macs/{}/vpc-id", mac)),
        );
        debug!("Found Instance AWS Metadata: {:?}", results);
        results
            .into_iter()
            .fold(HashMap::new(), |mut filtered, (prop, value)| {
                if let Some(value) = value {
                    filtered.insert(prop, value);
                }
                filtered
            })
    }

    fn lookup_metadata_key(&self, key: &str) -> Option<String> {
        let mut response = self
            .client
            .get(&format!("http://{}/latest/meta-data/{}", self.host, key))
            .send()
            .and_then(Response::error_for_status)
            .map_err(|e| {
                error!("Error requesting metadata key: {}", e);
                e
            })
            .ok()?;
        response.text().ok()
    }

    fn lookup_instance_identity(&self) -> Option<HashMap<String, Value>> {
        let mut response = self
            .client
            .get(&format!(
                "http://{}/latest/dynamic/instance-identity/document",
                self.host
            ))
            .send()
            .and_then(Response::error_for_status)
            .map_err(|e| {
                error!("Error requesting instance identity document: {}", e);
                e
            })
            .ok()?;
        response.json().ok()
    }
}


src/lib.rs
#![allow(unused, deprecated)]

extern crate itertools;
#[macro_use]
extern crate log;
extern crate percent_encoding;
#[macro_use]
extern crate quick_error;
extern crate rand;
extern crate reqwest;
extern crate serde;
#[macro_use]
extern crate serde_derive;
extern crate serde_json;

use reqwest::header::HeaderMap;
use reqwest::Client as ReqwestClient;
pub use reqwest::{Error as ReqwestError, Method, Response, StatusCode};
pub use serde::de::DeserializeOwned;
pub use serde::Serialize;

use self::instance::InstanceClient;
pub use self::instance::{Instance, PortData, SecurePort, StatusType};
use self::registry::RegistryClient;

mod aws;
mod instance;
mod registry;
mod resolver;
mod rest;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EurekaConfig {
    pub host: String,
    pub port: u16,
    pub heartbeat_interval: usize,
    pub registry_fetch_interval: usize,
    pub max_retries: usize,
    pub request_retry_delay: usize,
    pub fetch_registry: bool,
    pub filter_up_instances: bool,
    pub service_path: String,
    pub ssl: bool,
    pub use_dns: bool,
    pub prefer_same_zone: bool,
    pub cluster_refresh_interval: usize,
    pub fetch_metadata: bool,
    pub register_with_eureka: bool,
    pub use_local_metadata: bool,
    pub prefer_ip_address: bool,
}

impl Default for EurekaConfig {
    fn default() -> Self {
        EurekaConfig {
            host: "localhost".to_string(),
            port: 8761,
            heartbeat_interval: 30_000,
            registry_fetch_interval: 30_000,
            max_retries: 3,
            request_retry_delay: 500,
            fetch_registry: true,
            filter_up_instances: true,
            service_path: "/eureka".to_string(),
            ssl: false,
            use_dns: false,
            prefer_same_zone: true,
            cluster_refresh_interval: 300_000,
            fetch_metadata: true,
            register_with_eureka: true,
            use_local_metadata: false,
            prefer_ip_address: false,
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct BaseConfig {
    pub eureka: EurekaConfig,
    pub instance: Instance,
}

quick_error! {
    #[derive(Debug)]
    pub enum EurekaError {
        Network(err: ReqwestError) {
            description(err.description())
            cause(err)
        }
        Request(status: StatusCode) {
            description(status.canonical_reason().unwrap_or("Unknown Status Code"))
        }
        UnexpectedState(description: String) {
            description(description)
        }
        ParseError(description: String) {}
    }
}

#[derive(Debug)]
pub struct EurekaClient {
    base_url: String,
    config: BaseConfig,
    client: ReqwestClient,
    registry: RegistryClient,
    instance: Option<InstanceClient>,
}

impl EurekaClient {
    pub fn new(config: BaseConfig) -> Self {
        let base_url = {
            let ssl = config.eureka.ssl;
            let protocol = if ssl { "https" } else { "http" };
            let host = &config.eureka.host;
            let port = config.eureka.port;
            let service_path = &config.eureka.service_path;
            format!("{}://{}:{}{}", protocol, host, port, service_path)
        };
        let mut instance = config.instance.clone();
        instance.vip_address = instance.app.clone();
        instance.secure_vip_address = instance.vip_address.clone();
        EurekaClient {
            base_url: base_url.clone(),
            client: ReqwestClient::new(),
            registry: RegistryClient::new(base_url.clone()),
            instance: if config.eureka.register_with_eureka {
                Some(InstanceClient::new(base_url, instance))
            } else {
                None
            },
            config,
        }
    }

    pub fn start(&self) {
        self.registry.start();
        if let Some(ref instance) = self.instance {
            instance.start();
        }
    }

    pub fn find_app_address(&self, app_id: &str) -> Option<String> {
        let instance = self.registry.get_instance_by_app_name(app_id);
        if let Some(instance) = instance {
            let ssl = self.config.eureka.ssl;
            let host = instance.ip_addr;
            let port = if ssl {
                instance.secure_port.value
            } else {
                instance.port.value
            };
            println!("app {} addr {}:{}", app_id, host, port);
            Some(format!("{}:{}", host, port))
        } else {
            None
        }
    }

    /// Sends a request to another app in this eureka cluster, and returns the response.
    ///
    /// This method assumes that your services all communicate using JSON.
    /// Future methods may be added to allow other request body types.
    ///
    /// You can add additional headers such as `Authorization` using the `headers` parameter.
    pub fn make_request<V: Serialize>(
        &self,
        app: &str,
        path: &str,
        method: Method,
        body: &V,
        mut headers: HeaderMap,
    ) -> Result<Response, EurekaError> {
        log::debug!("finding app {}", app);
        let instance = self.registry.get_instance_by_app_name(app);
        if let Some(instance) = instance {
            //println!("app {} instance {:?}", app, instance);
            let ssl = self.config.eureka.ssl;
            let protocol = if ssl { "https" } else { "http" };
            let host = instance.ip_addr;
            let port = if ssl {
                instance.secure_port.value
            } else {
                instance.port.value
            };
            log::debug!("app {} addr {}:{}", app, host, port);
            self.client
                .request(
                    method,
                    &format!(
                        "{}://{}:{}/{}",
                        protocol,
                        host,
                        port,
                        path.trim_left_matches('/')
                    ),
                )
                .headers(headers)
                .json(body)
                .send()
                .map_err(EurekaError::Network)
        } else {
            Err(EurekaError::UnexpectedState(format!(
                "Could not find app {}",
                app
            )))
        }
    }

    pub fn call<V: Serialize, R: DeserializeOwned>(
        &self,
        app: &str,
        path: &str,
        method: Method,
        body: &V,
        mut headers: HeaderMap,
    ) -> Result<R, EurekaError> {
        let mut resp = self.make_request(app, path, method, body, headers)?;
        match resp.status() {
            StatusCode::OK => Ok(resp.json().map_err(EurekaError::Network)?),
            s => Err(EurekaError::Request(resp.status())),
        }
    }
}

fn path_segment_encode(value: &str) -> String {
    percent_encoding::utf8_percent_encode(value, percent_encoding::PATH_SEGMENT_ENCODE_SET)
        .to_string()
}

fn query_encode(value: &str) -> String {
    percent_encoding::utf8_percent_encode(value, percent_encoding::QUERY_ENCODE_SET).to_string()
}


src/instance.rs
pub use crate::rest::structures::{Instance, PortData, SecurePort, StatusType};
use crate::rest::EurekaRestClient;
use crate::EurekaError;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

#[derive(Debug)]
pub struct InstanceClient {
    client: Arc<EurekaRestClient>,
    config: Arc<Instance>,
    is_running: Arc<AtomicBool>,
}

impl InstanceClient {
    pub fn new(base_url: String, config: Instance) -> Self {
        InstanceClient {
            client: Arc::new(EurekaRestClient::new(base_url)),
            config: Arc::new(config),
            is_running: Arc::new(AtomicBool::new(false)),
        }
    }

    fn get_instance_id(&self) -> String {
        let mut instance_id = self.config.host_name.clone();
        if let Some(ref inst_id) = self.config.instance_id {
            instance_id = inst_id.clone();
        }
        instance_id
    }

    pub fn start(&self) {
        while let Err(e) = self.client.register(&self.config.app, &*self.config) {
            error!("Failed to register app: {}", e);
            thread::sleep(Duration::from_secs(15));
        }
        debug!("Registered app with eureka");

        self.is_running.store(true, Ordering::Relaxed);

        let is_running = Arc::clone(&self.is_running);
        let client = Arc::clone(&self.client);
        let config = Arc::clone(&self.config);
        let instance_id = self.get_instance_id();
        thread::spawn(move || {
            let do_regist = || {
                match client.register(&config.app, &*config) {
                    Ok(_) => {
                        info!("update status to UP");
                        client
                            .update_status(&config.app, &instance_id, StatusType::Up)
                            .map(|_| info!("update status ok!"))
                            .map_err(|err| error!("Failed to set app to UP: {}", err));
                    }
                    Err(e) => {
                        error!("Failed to register app: {}", e);
                    }
                };
            };
            thread::sleep(Duration::from_secs(30));
            while is_running.load(Ordering::Relaxed) {
                let resp = client.send_heartbeat(&config.app, &instance_id);
                match resp {
                    Err(EurekaError::UnexpectedState(_)) => {
                        warn!("App not registered with eureka, reregistering");
                        do_regist();
                    }
                    Err(e) => {
                        error!("Failed to send heartbeat: {}, reregistering", e);
                        do_regist();
                    }
                    Ok(_) => {
                        debug!("Sent heartbeat successfully");
                    }
                }
                thread::sleep(Duration::from_secs(30));
            }
        });

        while let Err(e) =
            self.client
                .update_status(&self.config.app, &self.get_instance_id(), StatusType::Up)
        {
            error!("Failed to set app to UP: {}", e);
            thread::sleep(Duration::from_secs(15));
        }
    }
}

impl Drop for InstanceClient {
    fn drop(&mut self) {
        self.is_running.store(false, Ordering::Relaxed);
        let _ = self
            .client
            .deregister(&self.config.app, &self.get_instance_id());
    }
}


src/registry.rs
use itertools::Itertools;
use rand::random;
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, RwLock};
use std::thread;
use std::time::Duration;

use crate::rest::structures::{Instance, StatusType};
use crate::rest::EurekaRestClient;

#[derive(Debug)]
pub struct RegistryClient {
    client: Arc<EurekaRestClient>,
    app_cache: Arc<RwLock<HashMap<String, Vec<Instance>>>>,
    is_running: Arc<AtomicBool>,
}

impl RegistryClient {
    pub fn new(base_url: String) -> Self {
        RegistryClient {
            client: Arc::new(EurekaRestClient::new(base_url)),
            app_cache: Arc::new(RwLock::new(HashMap::new())),
            is_running: Arc::new(AtomicBool::new(false)),
        }
    }

    pub fn update_app_cache(&self) -> Result<(), String> {
        RegistryClient::update_app_cache_internal(&self.client, &self.app_cache)
    }

    fn update_app_cache_internal(
        client: &Arc<EurekaRestClient>,
        app_cache: &Arc<RwLock<HashMap<String, Vec<Instance>>>>,
    ) -> Result<(), String> {
        let resp = client.get_all_instances();
        match resp {
            Ok(instances) => {
                // println!("got instances {:?}", instances);
                *app_cache.write().unwrap() = group_instances_by_app(instances);
                Ok(())
            }
            Err(e) => Err(format!("Failed to fetch registry: {:?}", e)),
        }
    }
    pub fn start(&self) {
        self.is_running.store(true, Ordering::Relaxed);

        let is_running = Arc::clone(&self.is_running);
        let client = Arc::clone(&self.client);
        let app_cache = Arc::clone(&self.app_cache);
        self.update_app_cache();
        thread::spawn(move || {
            while is_running.load(Ordering::Relaxed) {
                RegistryClient::update_app_cache_internal(&client, &app_cache)
                    .map_err(|e| println!("{}", e));
                thread::sleep(Duration::from_secs(30));
            }
        });
    }

    pub fn get_instance_by_app_name(&self, app: &str) -> Option<Instance> {
        // Clone the result to avoid holding onto a lock on the app cache indefinitely
        self.app_cache
            .read()
            .unwrap()
            .get(app)
            .and_then(|instances| {
                //random select one UP node
                let mut valid_ids: Vec<usize> = Vec::new();
                for (i, item) in instances.iter().enumerate() {
                    if item.status == StatusType::Up {
                        valid_ids.push(i);
                    }
                }
                if !valid_ids.is_empty() {
                    let index = valid_ids[random::<usize>() % valid_ids.len()];
                    instances.get(index)
                } else {
                    None
                }
            })
            .cloned()
    }
}

impl Drop for RegistryClient {
    fn drop(&mut self) {
        self.is_running.store(false, Ordering::Relaxed);
    }
}

fn group_instances_by_app(instances: Vec<Instance>) -> HashMap<String, Vec<Instance>> {
    instances
        .into_iter()
        .group_by(|i| i.app.clone())
        .into_iter()
        .map(|(k, g)| (k, g.collect()))
        .collect()
}


src/rest/structures.rs
use std::collections::HashMap;
use std::fmt;
use std::fmt::Display;
use std::fs::read;
use std::io::Write;
use std::net::ToSocketAddrs;
use std::str::FromStr;

use strong_xml::xmlparser::{ElementEnd, Token};
use strong_xml::{XmlRead, XmlReader, XmlResult, XmlWrite, XmlWriter};

#[derive(XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "applications")]
pub struct Applications {
    #[xml(flatten_text = "versions__delta")]
    pub versions_delta: Option<String>,
    #[xml(flatten_text = "apps__hashcode")]
    pub apps_hashcode: Option<String>,
    #[xml(child = "application")]
    pub applications: Vec<Application>,
}

#[derive(XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "application")]
pub struct Application {
    #[xml(flatten_text = "name")]
    pub name: String,
    #[xml(child = "instance")]
    pub instances: Vec<Instance>,
}

#[derive(Clone, XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "instance")]
pub struct Instance {
    #[xml(flatten_text = "hostName")]
    pub host_name: String,
    #[xml(flatten_text = "instanceId")]
    pub instance_id: Option<String>,
    #[xml(flatten_text = "app")]
    pub app: String,
    #[xml(flatten_text = "ipAddr")]
    pub ip_addr: String,
    #[xml(flatten_text = "vipAddress")]
    pub vip_address: String,
    #[xml(flatten_text = "secureVipAddress")]
    pub secure_vip_address: String,
    #[xml(flatten_text = "status")]
    pub status: StatusType,
    #[xml(child = "port")]
    pub port: PortData,
    #[xml(child = "securePort")]
    pub secure_port: SecurePort,
    #[xml(flatten_text = "homePageUrl")]
    pub home_page_url: String,
    #[xml(flatten_text = "statusPageUrl")]
    pub status_page_url: String,
    #[xml(flatten_text = "healthCheckUrl")]
    pub health_check_url: String,
    #[xml(child = "dataCenterInfo")]
    pub data_center_info: DataCenterInfo,
    #[xml(child = "leaseInfo")]
    pub lease_info: Option<LeaseInfo>,
    #[xml(child = "metadata")]
    pub metadata: Option<AppMetaDataType>,
}

impl Default for Instance {
    fn default() -> Self {
        Instance {
            host_name: "localhost".to_string(),
            instance_id: None,
            app: env!("CARGO_PKG_NAME").to_string(),
            ip_addr: "127.0.0.1".to_string(),
            vip_address: env!("CARGO_PKG_NAME").to_string(),
            secure_vip_address: env!("CARGO_PKG_NAME").to_string(),
            status: StatusType::Starting,
            port: PortData::default(),
            secure_port: SecurePort::default(),
            home_page_url: "".to_string(),
            status_page_url: "".to_string(),
            health_check_url: "".to_string(),
            data_center_info: DataCenterInfo::default(),
            lease_info: None,
            metadata: None,
        }
    }
}

#[derive(Clone, XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "port")]
pub struct PortData {
    #[xml(attr = "enabled")]
    pub enabled: bool,
    #[xml(default, text)]
    pub value: u16,
}

impl Default for PortData {
    fn default() -> Self {
        PortData {
            enabled: true,
            value: 80,
        }
    }
}

impl PortData {
    pub fn new(port: u16, enabled: bool) -> Self {
        Self {
            value: port,
            enabled,
        }
    }
}

#[derive(Clone, XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "securePort")]
pub struct SecurePort {
    #[xml(attr = "enabled")]
    pub enabled: bool,
    #[xml(default, text)]
    pub value: u16,
}

impl Default for SecurePort {
    fn default() -> Self {
        SecurePort {
            enabled: false,
            value: 443,
        }
    }
}

impl SecurePort {
    pub fn new(port: u16, enabled: bool) -> Self {
        Self {
            value: port,
            enabled,
        }
    }
}

#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum DcNameType {
    MyOwn,
    Amazon,
}

impl Display for DcNameType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Self::MyOwn => write!(f, "MyOwn"),
            Self::Amazon => write!(f, "Amazon"),
        }
    }
}

impl FromStr for DcNameType {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "MyOwn" => Ok(Self::MyOwn),
            "Amazon" => Ok(Self::Amazon),
            _ => Err("Invalid dcNameType".to_string()),
        }
    }
}

#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum StatusType {
    Up,
    Down,
    Starting,
    OutOfService,
    Unknown,
}

impl Display for StatusType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Self::Up => write!(f, "UP"),
            Self::Down => write!(f, "DOWN"),
            Self::Starting => write!(f, "STARTING"),
            Self::OutOfService => write!(f, "OUT_OF_SERVICE"),
            Self::Unknown => write!(f, "UNKNOWN"),
        }
    }
}

impl FromStr for StatusType {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "UP" => Ok(Self::Up),
            "DOWN" => Ok(Self::Down),
            "OUT_OF_SERVICE" => Ok(Self::OutOfService),
            "UNKNOWN" => Ok(Self::Unknown),
            _ => Err("Invalid statusType".to_string()),
        }
    }
}

#[derive(Clone, XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "metadata")]
pub struct AmazonMetaDataType {
    #[xml(flatten_text = "ami-launch-index")]
    pub ami_launch_index: String,
    #[xml(flatten_text = "local-hostname")]
    pub local_hostname: String,
    #[xml(flatten_text = "availability-zone")]
    pub availability_zone: String,
    #[xml(flatten_text = "instance-id")]
    pub instance_id: String,
    #[xml(flatten_text = "public-ipv4")]
    pub public_ipv4: String,
    #[xml(flatten_text = "public-hostname")]
    pub public_hostname: String,
    #[xml(flatten_text = "ami-manifest-patch")]
    pub ami_manifest_patch: String,
    #[xml(flatten_text = "local-ipv4")]
    pub local_ipv4: String,
    #[xml(flatten_text = "hostname")]
    pub hostname: String,
    #[xml(flatten_text = "ami-id")]
    pub ami_id: String,
    #[xml(flatten_text = "instance-type")]
    pub instance_type: String,
}

#[derive(Clone, XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "dataCenterInfo")]
pub struct DataCenterInfo {
    #[xml(attr = "class")]
    pub class: Option<String>,
    #[xml(flatten_text = "name")]
    pub name: DcNameType,
    #[xml(child = "metadata")]
    pub metadata: Option<AmazonMetaDataType>,
}

impl Default for DataCenterInfo {
    fn default() -> Self {
        DataCenterInfo {
            class: None,
            name: DcNameType::MyOwn,
            metadata: None,
        }
    }
}

#[derive(Clone, XmlWrite, XmlRead, PartialEq, Debug)]
#[xml(tag = "leaseInfo")]
pub struct LeaseInfo {
    #[xml(flatten_text = "evictionDurationInSecs")]
    pub eviction_duration_in_secs: Option<usize>,
}

#[derive(Clone, Default, Debug, PartialEq)]
pub struct AppMetaDataType {
    pub class: Option<String>,
    pub map: HashMap<String, String>,
}

impl AppMetaDataType {
    pub const TAG: &'static str = "metadata";
}

impl<'a> XmlRead<'a> for AppMetaDataType {
    fn from_reader(reader: &mut XmlReader<'a>) -> XmlResult<Self> {
        let mut metadata = Self::default();
        reader.read_till_element_start(AppMetaDataType::TAG)?;

        // attr class
        while let Some((attr_name, attr_val)) = reader.find_attribute()? {
            if attr_name.eq("class") {
                metadata.class = Some(attr_val.to_string());
            }
        }

        // end?
        let next = reader.next().unwrap()?;
        if let Token::ElementEnd {
            end: ElementEnd::Empty,
            ..
        } = next
        {
            return Ok(metadata);
        }

        // child
        while let Some(child_key) = reader.find_element_start(Some(AppMetaDataType::TAG))? {
            reader.next();
            let child_value = reader.read_text(child_key)?;
            metadata
                .map
                .insert(child_key.to_string(), child_value.to_string());
        }

        Ok(metadata)
    }
}

impl XmlWrite for AppMetaDataType {
    fn to_writer<W: Write>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()> {
        writer.write_element_start(AppMetaDataType::TAG);
        if let Some(v) = &self.class {
            writer.write_attribute("class", v.as_str())?;
        }
        writer.write_element_end_open()?;
        for (k, v) in &self.map {
            writer.write_flatten_text(k.as_str(), v.as_str(), false);
        }
        writer.write_element_end_close(AppMetaDataType::TAG);
        Ok(())
    }
}

mod tests {
    use std::str::FromStr;

    use strong_xml::{XmlRead, XmlResult};

    use super::*;

    #[derive(XmlRead, PartialEq, Debug)]
    #[xml(tag = "root")]
    struct R {
        #[xml(attr = "bar")]
        bar: Option<String>,
        #[xml(attr = "baz")]
        baz: Option<usize>,
    }

    #[test]
    fn test_xml_r() -> XmlResult<()> {
        let xml = r#"<root bar="bar" baz="123"/>"#;
        let r = R::from_str(xml)?;
        Ok(())
    }

    #[test]
    fn test_xml_applications() -> XmlResult<()> {
        let xml = r#"<applications></applications>"#;
        let applications = Applications::from_str(xml)?;
        Ok(())
    }

    #[test]
    fn test_xml_application() -> XmlResult<()> {
        let xml = r#"<application><name>abcd</name></application>"#;
        let application = Application::from_str(xml)?;
        Ok(())
    }

    #[test]
    fn test_xml_instance() -> XmlResult<()> {
        let xml = r#"<instance>
      <hostName>localhost</hostName>
      <app>BENCH</app>
      <ipAddr>127.0.0.1</ipAddr>
      <status>UP</status>
      <overriddenstatus>UP</overriddenstatus>
      <port enabled="true">8080</port>
      <securePort enabled="false">443</securePort>
      <countryId>1</countryId>
      <dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
        <name>MyOwn</name>
      </dataCenterInfo>
      <leaseInfo>
        <renewalIntervalInSecs>30</renewalIntervalInSecs>
        <durationInSecs>90</durationInSecs>
        <registrationTimestamp>1616761261538</registrationTimestamp>
        <lastRenewalTimestamp>1616761921820</lastRenewalTimestamp>
        <evictionTimestamp>0</evictionTimestamp>
        <serviceUpTimestamp>1616761261439</serviceUpTimestamp>
      </leaseInfo>
      <metadata class="java.util.Collections$EmptyMap"/>
      <homePageUrl>/eureka</homePageUrl>
      <statusPageUrl></statusPageUrl>
      <healthCheckUrl></healthCheckUrl>
      <vipAddress>bench</vipAddress>
      <secureVipAddress>bench</secureVipAddress>
      <isCoordinatingDiscoveryServer>false</isCoordinatingDiscoveryServer>
      <lastUpdatedTimestamp>1616761261538</lastUpdatedTimestamp>
      <lastDirtyTimestamp>1616761261439</lastDirtyTimestamp>
      <actionType>ADDED</actionType>
    </instance>"#;

        let instance: Instance = Instance::from_str(xml)?;
        assert_eq!(instance.host_name, "localhost");
        assert_eq!(instance.app, "BENCH");
        assert_eq!(instance.ip_addr, "127.0.0.1");

        Ok(())
    }

    #[test]
    fn test_xml_port() -> XmlResult<()> {
        let xml = r#"<port enabled="false">80</port>"#;
        let port = PortData::from_str(xml)?;
        assert_eq!(port.enabled, false);
        assert_eq!(port.value, 80);
        Ok(())
    }

    #[test]
    fn test_xml_data_center_info() -> XmlResult<()> {
        let xml = r#"<dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
        <name>MyOwn</name>
      </dataCenterInfo>"#;
        let data_center_info = DataCenterInfo::from_str(xml)?;
        assert_eq!(data_center_info.name, DcNameType::MyOwn);
        assert_eq!(
            data_center_info.class,
            Some("com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo".to_string())
        );

        Ok(())
    }

    #[test]
    fn test_xml_lease_info() -> XmlResult<()> {
        let xml = r#"<leaseInfo></leaseInfo>"#;
        let lease_info = LeaseInfo::from_str(xml)?;
        assert_eq!(lease_info.eviction_duration_in_secs, None);
        Ok(())
    }

    #[test]
    fn test_xml_app_meta_data() -> XmlResult<()> {
        let xml = r#"<metadata class="java.util.Collections$EmptyMap"><a>hello</a></metadata>"#;
        let metadata = AppMetaDataType::from_str(xml)?;
        assert_eq!(
            metadata.class,
            Some("java.util.Collections$EmptyMap".to_string())
        );
        assert_eq!(metadata.map.get("a").unwrap(), "hello");

        let s = metadata.to_string()?;
        assert_eq!(s, xml);

        Ok(())
    }

    #[test]
    fn test_xml_full() -> XmlResult<()> {
        let xml = r#"<applications>
  <versions__delta>1</versions__delta>
  <apps__hashcode>UP_2_</apps__hashcode>
  <application>
    <name>BENCH</name>
    <instance>
      <hostName>localhost</hostName>
      <app>BENCH</app>
      <ipAddr>127.0.0.1</ipAddr>
      <status>UP</status>
      <overriddenstatus>UP</overriddenstatus>
      <port enabled="true">8080</port>
      <securePort enabled="false">443</securePort>
      <countryId>1</countryId>
      <dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
        <name>MyOwn</name>
      </dataCenterInfo>
      <leaseInfo>
        <renewalIntervalInSecs>30</renewalIntervalInSecs>
        <durationInSecs>90</durationInSecs>
        <registrationTimestamp>1616761261538</registrationTimestamp>
        <lastRenewalTimestamp>1616761921820</lastRenewalTimestamp>
        <evictionTimestamp>0</evictionTimestamp>
        <serviceUpTimestamp>1616761261439</serviceUpTimestamp>
      </leaseInfo>
      <metadata class="java.util.Collections$EmptyMap"/>
      <homePageUrl>/eureka</homePageUrl>
      <statusPageUrl></statusPageUrl>
      <healthCheckUrl></healthCheckUrl>
      <vipAddress>bench</vipAddress>
      <secureVipAddress>bench</secureVipAddress>
      <isCoordinatingDiscoveryServer>false</isCoordinatingDiscoveryServer>
      <lastUpdatedTimestamp>1616761261538</lastUpdatedTimestamp>
      <lastDirtyTimestamp>1616761261439</lastDirtyTimestamp>
      <actionType>ADDED</actionType>
    </instance>
    <instance>
      <hostName>localhost2</hostName>
      <app>BENCH</app>
      <ipAddr>127.0.0.1</ipAddr>
      <status>UP</status>
      <overriddenstatus>UP</overriddenstatus>
      <port enabled="true">8081</port>
      <securePort enabled="false">443</securePort>
      <countryId>1</countryId>
      <dataCenterInfo class="com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo">
        <name>MyOwn</name>
      </dataCenterInfo>
      <leaseInfo>
        <renewalIntervalInSecs>30</renewalIntervalInSecs>
        <durationInSecs>90</durationInSecs>
        <registrationTimestamp>1616761233029</registrationTimestamp>
        <lastRenewalTimestamp>1616761900818</lastRenewalTimestamp>
        <evictionTimestamp>0</evictionTimestamp>
        <serviceUpTimestamp>1616761232774</serviceUpTimestamp>
      </leaseInfo>
      <metadata class="java.util.Collections$EmptyMap"/>
      <homePageUrl>/eureka</homePageUrl>
      <statusPageUrl></statusPageUrl>
      <healthCheckUrl></healthCheckUrl>
      <vipAddress>bench</vipAddress>
      <secureVipAddress>bench</secureVipAddress>
      <isCoordinatingDiscoveryServer>false</isCoordinatingDiscoveryServer>
      <lastUpdatedTimestamp>1616761233030</lastUpdatedTimestamp>
      <lastDirtyTimestamp>1616761232774</lastDirtyTimestamp>
      <actionType>ADDED</actionType>
    </instance>
  </application>
</applications>"#;

        let application = Applications::from_str(xml)?;
        Ok(())
    }
}
